# Ebay used cars Germany

The aim of this project is to practice datacleaning. The dataset before us has been downloaded from Kaggle. On Kaggle the dataset is a lot cleaner than here, so first we are going to clean and then move onto the analysis phase. 

In [1]:
#Firstly we import our libaries
import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## first glance
For now we can see a few things which should be cleaned
* The titles have strange uppercases.
* The price has a dollarsign and a comma. (Very strange to see a dollarsign in a German website imho).
* The odometer column should be converted to an int64 column by removing the text.
* The null values should be inspected.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest','vehicle_type', 'registration_year', 'gearbox', 'powerps', 'model','odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
#assigning new column names to the dataset.

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerps   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage         date_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Changes in titles
In order to make the titles more workable I have converted them to snakecase. Moreover, some columns had names which were quite illogical like notRepairedDamage. Hence they have been adapted into something more stylish.

In [8]:
autos.describe()

registration_year       powerps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

### Numerics
Let us investigate the numeric aspect of the cleaning:
* nr_of_pictures has values of 0.0 only, hence that column is not interesting and will be deleted. 
* seller has the same values and will be deleted as well.
* The max year is 9999 which is a bit odd.
* The horsepower(ps) max value is also skyhigh, I believe there had to be a rocket for sale ;)
* The columns that are not there which should are price and odometer. First we should clean the values there.

In [9]:
# Let us correct the price and odometer columns and convert it to int.
autos["price"] = autos["price"].str.replace("$","").str.replace(",","")
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","")

# now converting them to integers
autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].astype(int)

autos.rename(columns={"odometer":"odometer_km"}, inplace = True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  powerps   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage         date_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [10]:
autos = autos.drop(["seller","nr_of_pictures"], axis = 1)

In [11]:
autos["odometer_km"].unique().shape

(13,)

In [12]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [13]:
autos["odometer_km"].value_counts().sort_index(ascending= True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [14]:
autos["price"].unique().shape

(2357,)

In [15]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
autos["price"].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

In [17]:
autos = autos[autos["price"].between(200,350000)]

In [18]:
autos["price"].describe()

count     47645.000000
mean       6000.707273
std        9110.783444
min         200.000000
25%        1300.000000
50%        3190.000000
75%        7500.000000
max      350000.000000
Name: price, dtype: float64

### Removing outliers
In the above cells we have been looking for outlier values. For odometer, the values seemed to be completely normal. However, in the price category there are some strange values like 100,000,000 or 1,234,566. Also, there were a lot of cars for free, which is probably a sign that ebay users can bid on the car. Similar to 100,000,000. 

In the end I have decided to remove the prices below 200 Dollars and above 350,000 Dollars. Because there can very well be some expensive classic cars for example. There are actually Ferraris on Ebay!

If we look at the remaining values, we can see that 75% cars on Ebay are under 10,000 Dollars. And 25% is actually under 1,300 Dollars!

In [19]:
autos[['date_crawled','date_created','last_seen']][0:5]

date_crawled         date_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [20]:
# These values are too precise to work with, 
# Therefore we are removing the times using the first 10 characters of the string only.
autos["date_crawled"] = autos["date_crawled"].str[:10]
autos["date_created"] = autos["date_created"].str[:10]
autos["last_seen"] = autos["last_seen"].str[:10]

In [21]:
autos["date_crawled"].value_counts(normalize = True, dropna=False).sort_index()

2016-03-05    0.025354
2016-03-06    0.014062
2016-03-07    0.035995
2016-03-08    0.033120
2016-03-09    0.033036
2016-03-10    0.032322
2016-03-11    0.032700
2016-03-12    0.036877
2016-03-13    0.015699
2016-03-14    0.036562
2016-03-15    0.034232
2016-03-16    0.029447
2016-03-17    0.031546
2016-03-18    0.012824
2016-03-19    0.034610
2016-03-20    0.037800
2016-03-21    0.037360
2016-03-22    0.032700
2016-03-23    0.032385
2016-03-24    0.029321
2016-03-25    0.031420
2016-03-26    0.032217
2016-03-27    0.031210
2016-03-28    0.035030
2016-03-29    0.033980
2016-03-30    0.033897
2016-03-31    0.031840
2016-04-01    0.033813
2016-04-02    0.035681
2016-04-03    0.038724
2016-04-04    0.036562
2016-04-05    0.013139
2016-04-06    0.003169
2016-04-07    0.001364
Name: date_crawled, dtype: float64

### Date Crawled
This column describes the dates when the content was saved by the website itself. We can see that after three days most days have been crawled three percent. This shows that the website is doing this regularly. I could not spot anything out of the ordinary.

In [22]:
autos["date_created"].value_counts(normalize = True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000042
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000063
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000042
2016-02-05    0.000042
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000042
2016-02-14    0.000042
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000042
2016-02-19    0.000063
2016-02-20    0.000042
2016-02-21    0.000063
                ...   
2016-03-09    0.033141
2016-03-10    0.032029
2016-03-11    0.033015
2016-03-12    0.036688
2016-03-13    0.017106
2016-03-14    0.035177
2016-03-15    0.033959
2016-03-16    0.029951
2016-03-17    0.031210
2016-03-18    0.013454
2016-03-19    0.033519
2016-03-20    0.037884
2016-03-21 

### Date Created
We can observe that most of the ads were created in the two months before. But there are a handful dates from 6 months before. 

In [23]:
autos["last_seen"].value_counts(normalize = True, dropna=False).sort_index()

2016-03-05    0.001091
2016-03-06    0.004303
2016-03-07    0.005373
2016-03-08    0.007178
2016-03-09    0.009613
2016-03-10    0.010473
2016-03-11    0.012299
2016-03-12    0.023906
2016-03-13    0.008899
2016-03-14    0.012530
2016-03-15    0.015720
2016-03-16    0.016287
2016-03-17    0.028083
2016-03-18    0.007283
2016-03-19    0.015615
2016-03-20    0.020653
2016-03-21    0.020506
2016-03-22    0.021408
2016-03-23    0.018491
2016-03-24    0.019624
2016-03-25    0.019079
2016-03-26    0.016686
2016-03-27    0.015511
2016-03-28    0.020758
2016-03-29    0.022185
2016-03-30    0.024578
2016-03-31    0.023864
2016-04-01    0.022899
2016-04-02    0.024808
2016-04-03    0.025102
2016-04-04    0.024620
2016-04-05    0.125470
2016-04-06    0.222437
2016-04-07    0.132669
Name: last_seen, dtype: float64

### Last Seen
The results here are quite normal as well. With 13% seen 'today' 22% 'yesterday' and 12% the day before that. It is what we would expect from the behaviour of vendors.

In [24]:
autos["registration_year"].describe()

count    47645.000000
mean      2004.800084
std         88.423872
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Registration Year
Here we can see that there are some anomalies. Like the year 1000 and 9999. Those should be removed. Moreover, a standard deviation of 88 years for cars is quite a lot considering that most cars are not functioning anymore after 25 years.
* The highest value is clearly the 'now' in the table which is 2016.
* The lowest value could be very low, the car has been invented around 1900 and there could still be a car sold from that era. For now I will stick to 1900. This could be altered later by our findings.

In [25]:
autos["registration_year"].value_counts(normalize = True, dropna=False).sort_index()

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000042
1910    0.000042
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000084
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000021
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000084
1957    0.000042
1958    0.000084
1959    0.000126
1960    0.000441
1961    0.000126
1962    0.000084
1963    0.000168
1964    0.000252
          ...   
2000    0.063364
2001    0.054801
2002    0.051842
2003    0.056501
2004    0.056543
2005    0.060993
2006    0.055976
2007    0.047686
2008    0.046301
2009    0.043635
2010    0.033309
2011    0.033939
2012    0.027432
2013    0.016728
2014    0.013768
2015    0.007892
2016    0.024368
2017    0.028775
2018    0.009823
2019    0.000021
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000063
5911    0.000021
6200    0.000021
8888    0.0000

In [26]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize = True)

2000    0.065937
2005    0.063469
1999    0.062137
2004    0.058839
2003    0.058795
2006    0.058249
2001    0.057026
2002    0.053947
1998    0.049906
2007    0.049622
2008    0.048181
2009    0.045407
1997    0.040449
2011    0.035316
2010    0.034661
2012    0.028546
1996    0.028437
2016    0.025357
1995    0.024440
2013    0.017407
2014    0.014328
1994    0.012952
1993    0.008867
2015    0.008212
1992    0.007753
1991    0.007207
1990    0.006989
1989    0.003691
1988    0.002905
1985    0.001922
          ...   
1976    0.000459
1960    0.000459
1969    0.000415
1975    0.000393
1965    0.000371
1964    0.000262
1963    0.000175
1961    0.000131
1959    0.000131
1962    0.000087
1956    0.000087
1937    0.000087
1958    0.000087
1955    0.000044
1954    0.000044
1957    0.000044
1951    0.000044
1910    0.000044
1934    0.000044
1941    0.000044
1927    0.000022
1929    0.000022
1950    0.000022
1931    0.000022
1948    0.000022
1938    0.000022
1939    0.000022
1953    0.0000

### Registration Year
After removing the outside values, we can see that most of the cars listed have been built between 1998 and 2007. 

In [27]:
# Now let us find out which brands are the most common
autos["brand"].value_counts(normalize = 1)

volkswagen        0.211244
bmw               0.111257
opel              0.105906
mercedes_benz     0.097628
audi              0.087603
ford              0.068733
renault           0.046543
peugeot           0.029944
fiat              0.025095
seat              0.018171
skoda             0.016512
nissan            0.015289
mazda             0.015136
smart             0.014371
citroen           0.014131
toyota            0.012930
hyundai           0.010112
sonstige_autos    0.009544
volvo             0.009195
mini              0.008889
mitsubishi        0.008168
honda             0.007928
kia               0.007142
alfa_romeo        0.006683
porsche           0.006094
suzuki            0.005810
chevrolet         0.005722
chrysler          0.003560
dacia             0.002686
daihatsu          0.002490
jeep              0.002315
land_rover        0.002140
subaru            0.002097
saab              0.001660
jaguar            0.001529
daewoo            0.001485
rover             0.001332
t

## Brands mean price 
In order to have a large enough dataset for an accurate calculation of the mean price, I will select the brands that have at least 1% in the dataset. 500 entries should be enough to create a reliable value for mean price per brand.

In [28]:
autos["brand"].value_counts(normalize = 1).index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini',
       'mitsubishi', 'honda', 'kia', 'alfa_romeo', 'porsche', 'suzuki',
       'chevrolet', 'chrysler', 'dacia', 'daihatsu', 'jeep', 'land_rover',
       'subaru', 'saab', 'jaguar', 'daewoo', 'rover', 'trabant', 'lancia',
       'lada'],
      dtype='object')

In [29]:
# The brands that have more than 1% are up to hyundai
brands1pct = ['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai']

In [32]:
# let's aggegrate and create a dictionary which shows mean price
# for the brands1pct.
brands1pct_mean_price = {}

for b in brands1pct:
    brand = autos[autos["brand"] == b]
    mean_price = brand["price"].mean()
    mean_price = mean_price.round()
    brands1pct_mean_price[b] = mean_price
    
print(brands1pct_mean_price)

{'smart': 3596.0, 'skoda': 6451.0, 'renault': 2553.0, 'citroen': 3819.0, 'audi': 9406.0, 'fiat': 2926.0, 'bmw': 8403.0, 'mazda': 4205.0, 'peugeot': 3142.0, 'hyundai': 5422.0, 'nissan': 4829.0, 'seat': 4505.0, 'ford': 3883.0, 'volkswagen': 5506.0, 'toyota': 5176.0, 'opel': 3078.0, 'mercedes_benz': 8692.0}


## Mean prices per brand
The three most expensive popular brands on ebay are:
* 1 Audi - €9406
* 2 Mercedes-Benz - €8692
* 3 Bmw - €8403

The three cheapest brands are:
* 1 Renault - €2553
* 2 Fiat - €2926
* 3 Opel - €3078

In [33]:
# Right now we will make another dictionary in order to compare the
# mean price to the mean mileage.
brands1pct_mean_mileage = {}

for b in brands1pct:
    brand = autos[autos["brand"] == b]
    mean_mileage = brand["odometer_km"].mean()
    mean_mileage = mean_mileage.round()
    brands1pct_mean_mileage[b] = mean_mileage
    
print(brands1pct_mean_mileage)

{'smart': 99734.0, 'skoda': 110999.0, 'renault': 128053.0, 'citroen': 119815.0, 'audi': 129261.0, 'fiat': 116950.0, 'bmw': 132793.0, 'mazda': 124076.0, 'peugeot': 126929.0, 'hyundai': 106793.0, 'nissan': 118179.0, 'seat': 121605.0, 'ford': 124096.0, 'volkswagen': 128775.0, 'toyota': 116106.0, 'opel': 129232.0, 'mercedes_benz': 131092.0}


In [40]:
bmprice = pd.Series(brands1pct_mean_price)
print(bmprice)

audi             9406.0
bmw              8403.0
citroen          3819.0
fiat             2926.0
ford             3883.0
hyundai          5422.0
mazda            4205.0
mercedes_benz    8692.0
nissan           4829.0
opel             3078.0
peugeot          3142.0
renault          2553.0
seat             4505.0
skoda            6451.0
smart            3596.0
toyota           5176.0
volkswagen       5506.0
dtype: float64


In [41]:
bmprice = pd.DataFrame(bmprice, columns=["mean_price"])
bmprice

mean_price
audi               9406.0
bmw                8403.0
citroen            3819.0
fiat               2926.0
ford               3883.0
hyundai            5422.0
mazda              4205.0
mercedes_benz      8692.0
nissan             4829.0
opel               3078.0
peugeot            3142.0
renault            2553.0
seat               4505.0
skoda              6451.0
smart              3596.0
toyota             5176.0
volkswagen         5506.0

In [53]:
bmodometer = pd.Series(brands1pct_mean_mileage)
bmodometer

audi             129261.0
bmw              132793.0
citroen          119815.0
fiat             116950.0
ford             124096.0
hyundai          106793.0
mazda            124076.0
mercedes_benz    131092.0
nissan           118179.0
opel             129232.0
peugeot          126929.0
renault          128053.0
seat             121605.0
skoda            110999.0
smart             99734.0
toyota           116106.0
volkswagen       128775.0
dtype: float64

In [56]:
bmprice["mean_odometer_km"] = bmodometer
bmprice

mean_price  mean_odometer_km
audi               9406.0          129261.0
bmw                8403.0          132793.0
citroen            3819.0          119815.0
fiat               2926.0          116950.0
ford               3883.0          124096.0
hyundai            5422.0          106793.0
mazda              4205.0          124076.0
mercedes_benz      8692.0          131092.0
nissan             4829.0          118179.0
opel               3078.0          129232.0
peugeot            3142.0          126929.0
renault            2553.0          128053.0
seat               4505.0          121605.0
skoda              6451.0          110999.0
smart              3596.0           99734.0
toyota             5176.0          116106.0
volkswagen         5506.0          128775.0

## Mean odometer comparison
When we are comparing the mean odometer to the mean price we can see that the highest values are for:
* 1 BMW - 132793km
* 2 Mercedes-Benz - 131092km
* 3 Audi - 129261km

It means that even though these expensive cars drive a long way, they do not lose as much of their value as the others do. 

In the cheap sector we can see that the best buy would be to go for a Fiat. They have a low mileage and a low price. 